text/000x-nrd-kernels.md

```diff
@@ -0,0 +1,248 @@
+
+Each node maintains an index of _recent_ NRD kernels to enable efficient checking of NRD relative lock heights. Note we only need to index NRD locks and we only need to index those within recent history. Relative locks longer than 7 days are not valid. This is believed to be sufficient to cover all proposed use cases.
+
+The minimum value for a relative lock height is 1 meaning a prior instance of the kernel can exist in the previous block for the lock criteria to be met.
+An instance of the NRD kernel in the _same_ block will invalidate the block as the lock criteria will not be met.
+
+NRD lock heights of 0 are invalid and it is never valid for two duplicate instances of the _same_ NRD kernel to exist in the same block.
+
+It follows that two transactions contaning duplicate instances of the same NRD kernel cannot be accepted as valid in the txpool concurrently.
+"First one wins" semantics apply when validating transactions containing NRD kernels in a similar way to resolving the spending of unspent outputs.
+
+Grin supports "rewind" back through recent history to handle fork and chain reorg scenarios. 1 week of full blocks are maintained on each node and up to 10080 blocks can be rewound. To support relative lockheights each node must maintain an index over sufficient kernel history for an _additional_ 10080 blocks beyond this rewind horizon. Each node should maintain 2 weeks of kernel history in the local NRD kernel index. This will cover the pathological case of a 1 week rewind and the validaton of a 1 week long relative lock beyond that. The primary use case is for revocable payment channel close operations. We believe a 7 day period is more than sufficient for this. We do not require long, extended revocation periods and limiting this to a few days is preferable to keep the cost of verification low. The need for these revocable transactions to be included on chain should be low as these are only required in a non-cooperative situation but where required we want to minimize the cost of verification which must be performed across all nodes.
+
+----
+
+The following kernel variants are supported in Grin -
+
+* Plain
+* Coinbase
+* HeightLocked
+* NoRecentDuplicate
+
+These are implemented as kernel "feature" variants -
+
+```
+pub enum KernelFeatures {
+	/// Plain kernel (default for Grin txs).
+	Plain = 0,
+	/// A coinbase kernel.
+	Coinbase = 1,
+	/// A kernel with an explicit absolute lock height.
+	HeightLocked = 2,
+	/// A relative height locked NRD kernel.
+	NoRecentDuplicate = 3,
+}
+```
+
+Each kernel variant includes feature specific data -
+
+```
+# Plain
+{    
+  "fee": 8
+}
+# Coinbase
+{
+  # empty
+}
+# Height Locked
+{
+  "fee": 8,
+  "lock_height": 295800
+}
+# No Recent Duplicate (NRD)
+{
+  "fee": 8,
+  "relative_height": 1440,
+}
+```
```

Comment:
Note that method kernel_sig_msg() would be extended with a clause

			KernelFeatures:: NoRecentDuplicate{ fee, relative_height } => (x, fee, relative_height).hash(),

Reply:
Good point. I did not touch on the kernel sig msg at all in this RFC. I'll add a section for this.